---
title: "Get annotated, differentially-methylated regions by comparing samples from two conditions"
author: "Tsunghan Hsieh"
description: >
  This file demonstrates how to get annotated DMR.
date: "`r format(Sys.Date(), format='%d-%m-%Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Clustering_Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
# Set RMarkDown options
knitr::opts_chunk$set(echo = TRUE, warning = TRUE, message = TRUE, eval = TRUE, collapse = FALSE)

# Turn off scientific notation
options(scipen = 999)

# load all functions
devtools::document()
devtools::load_all()
```

# Get adjusted methylation counts
***
We first get the methylation counts for each sample.

```{r todo}
# Run the following Rmd file:
# ./Doc/getUnadjustedMethylationCount2018.rmd.
# ./Doc/getUnadjustedMethylationCount2022.rmd.
```

# Set-up
***

## Load libraries
```{r library}
library(dplyr)
library(methylKit)
library(here)
library(stringr)
```

## Set dir
***
Set input and output dir.

```{r dir}
sampleSheetDir <- here::here("Doc","sampleSheet")
methylCountDir <- here::here("Results", "Statistics")
dmrDir <- here::here("Results", "Statistics", "DMR")
annotatedDmrDir <- here::here("Results", "Statistics", "DMR_annotated")
```

# Sample Sheet
***
```{r sampleSheet}
sampleSheet_SI <- read.csv(here::here(
  sampleSheetDir, "RRBS_2018_SI_sampleSheet.csv"))
sampleSheet_SI <- sampleSheet_SI[,2:ncol(sampleSheet_SI)]
sampleSheet_SI <- sampleSheet_SI |>
  dplyr::mutate(sample = paste0(Tissue, "_", Condition, "_", Age_w, "_", Replicate))

sampleSheet_LI <- read.csv(here::here(
  sampleSheetDir, "RRBS_2022_LI_sampleSheet.csv"))
sampleSheet_LI <- sampleSheet_LI[,2:ncol(sampleSheet_LI)]
sampleSheet_LI <- sampleSheet_LI |>
  dplyr::mutate(sample = paste0(Tissue, "_", Condition, "_", Age_w, "_", Replicate))
```

# TODO: Get DMR
***
Get DMR for 5hmc + 5mc, 5hmc, and 5mc. For more information, check `./Doc/getDMR_2018.Rmd` and `getDMR_2022.Rmd`

```{r DMR}
# Get DMR
```

# Methyl counts analysis
## Get methyl counts
***

```{r methyl_count}
# List all files in the folder with names containing "all"
methyl_counts_file_list <- list.files(
  path = here::here(methylCountDir, "5hmc_5mc"))
methyl_counts_list <- list()
for (file in methyl_counts_file_list) {
  count_fname_index <- unlist(strsplit(file, "_"))
  tissue <- count_fname_index[1]
  condition <- count_fname_index[2]
  weeks <- count_fname_index[3]
  replicate <- count_fname_index[5]
  
  new_index <- paste0(tissue, "_", 
                      condition, "_",
                      weeks, "_",
                      replicate)
  df <- read.csv(here::here(methylCountDir, "5hmc_5mc", file))
  df <- df |>
    dplyr::mutate(pos_index = paste0(chr, "_", start, "_", end)) |>
    dplyr::select(pos_index, freqC)
  
  colnames(df) <- c("pos_index", new_index)
  
  methyl_counts_list[[new_index]] <- df
}
```

## Overview analysis
***
Count the union methylated regions.

```{r methylcount_overview}
union_list <- list()
for (i in 1:(length(methyl_counts_list)-1)) {
  union_list <- union(union_list,
                      union(
                        methyl_counts_list[[i]]$pos_index,
                        methyl_counts_list[[i+1]]$pos_index))
}
message("There are total ", length(union_list), " methylated regions.")
```

## CpG 
***
Distribution of those methylated regions among CpG islands, CpG shores, and inter CpG regions

```{r methylated_cpg}
```

## Genes
***
Distribution of those methylated regions among promoter, exon, and intron regions.
```{r methylated_genes}
```

# DMR analysis
## Get annotated DMR
***
Get annotated DMR for 5hmc + 5mc, 5hmc, and 5mc.

```{r annotated_DMR, eval=FALSE}
dmrList <- list.files(dmrDir)

## Annotate CpG islands
for (file in dmrList) {
  outputfname <- sub(".*/([^.]+)\\.csv", "\\1", file)
  outputfname <- paste0("cpg_annotated_", outputfname)
  outputdir <- here::here("Results/Statistics/DMR_annotated/")
  output <- here::here(outputdir, outputfname)
  
  annotateDMR(
    input = here::here(dmrDir,file),
    col_names = c("chr", "start", "end", "strand"),
    annots = c("mm10_cpg_islands", "mm10_cpg_shores","mm10_cpg_shelves", "mm10_cpg_inter"),
    output = output)
}

## Annotate basic genes
## Could be longer than CpG annotation
for (file in dmrList) {
  outputfname <- sub(".*/([^.]+)\\.csv", "\\1", file)
  outputfname <- paste0("genes_annotated_", outputfname)
  outputdir <- here::here("Results/Statistics/DMR_annotated/")
  output <- here::here(outputdir, outputfname)
  
  annotateDMR(
    input = here::here(dmrDir,file),
    col_names = c("chr", "start", "end", "strand"),
    annots = c("mm10_basicgenes"),
    output = output)
}
```


## Overview analysis
***
Some over analysis for the total dmr

```{r overview_cpg}
# Read the annotated DMR
dmr_list <- list()
for (file in list.files(annotatedDmrDir)) {
  dmr_fname_index <- unlist(strsplit(unlist(strsplit(file, "\\."))[1], "_"))
  annotation <- dmr_fname_index[1]
  type <- dmr_fname_index[4]
  tissue <- dmr_fname_index[6]
  weeks <- dmr_fname_index[7]
  index <- paste0(annotation, "_", type, "_", tissue, "_", weeks)
  dmr_list[[index]] <- read.csv(here::here(annotatedDmrDir, file))
}

# Total number of unique DMR
union <- list()
for (i in 1:(length(dmr_list)-1)) {
  annotation <- unlist(strsplit(names(dmr_list[i]), "_"))[1]
  dmr_type <- unlist(strsplit(names(dmr_list[i]), "_"))[2]
  
  if(annotation == "cpg") {
    tmp_df1 <- dmr_list[[i]] |>
      dplyr::mutate(index = paste0(chr, "_", start, "_", end)) |>
      dplyr::select(index)
    tmp_df2 <- dmr_list[[i+1]] |>
      dplyr::mutate(index = paste0(chr, "_", start, "_", end)) |>
      dplyr::select(index)
    tmp_list <- union(tmp_df1$index, tmp_df2$index)
    union <- union(union, tmp_list)
  }
}
message("Total ", length(union), " DMRs were obtained.\n")

# Total number of unique 5mc DMR
union <- list()
for (i in 1:(length(dmr_list)-1)) {
  annotation <- unlist(strsplit(names(dmr_list[i]), "_"))[1]
  dmr_type <- unlist(strsplit(names(dmr_list[i]), "_"))[2]
  
  if (dmr_type == "5mc" && annotation == "cpg") {
    tmp_df1 <- dmr_list[[i]] |>
      dplyr::mutate(index = paste0(chr, "_", start, "_", end)) |>
      dplyr::select(index)
    tmp_df2 <- dmr_list[[i+1]] |>
      dplyr::mutate(index = paste0(chr, "_", start, "_", end)) |>
      dplyr::select(index)
    tmp_list <- union(tmp_df1$index, tmp_df2$index)
    union <- union(union, tmp_list)
  }
}
message("Total ", length(union), " 5mc DMRs were obtained.\n")

union <- list()
for (i in 1:(length(dmr_list)-1)) {
  annotation <- unlist(strsplit(names(dmr_list[i]), "_"))[1]
  dmr_type <- unlist(strsplit(names(dmr_list[i]), "_"))[2]
  
  if (dmr_type == "5hmc" && annotation == "cpg") {
    tmp_df1 <- dmr_list[[i]] |>
    dplyr::mutate(index = paste0(chr, "_", start, "_", end)) |>
    dplyr::select(index)
  tmp_df2 <- dmr_list[[i+1]] |>
    dplyr::mutate(index = paste0(chr, "_", start, "_", end)) |>
    dplyr::select(index)
  tmp_list <- union(tmp_df1$index, tmp_df2$index)
  union <- union(union, tmp_list)
  }
}
message("Total ", length(union), " 5hmc DMRs were obtained.\n")

```
